src/proof/er/er_proof.cpp

```diff
@@ -0,0 +1,378 @@
+  std::ofstream dratStream(dratFilename);
+  dratStream << dratBinary;
+  dratStream.close();
+
+  // Invoke drat2er
+#if CVC4_USE_DRAT2ER
+  drat2er::TransformDRATToExtendedResolution(formulaFilename,
+                                             dratFilename,
+                                             tracecheckFilename,
+                                             false,
+                                             drat2er::options::QUIET);
+#else
+  Unimplemented(
+      "ER proof production requires drat2er.\n"
+      "Run contrib/get-drat2er, reconfigure with --drat2er, and rebuild");
+#endif
+
+  // Parse the resulting TRACECHECK proof into an ER proof.
+  std::ifstream tracecheckStream(tracecheckFilename);
+  ErProof proof(usedClauses, TraceCheckProof::fromText(tracecheckStream));
+  tracecheckStream.close();
+
+  unlink(tracecheckFilename);
+  unlink(formulaFilename);
+  unlink(dratFilename);
+
+  return proof;
+}
+
+ErProof::ErProof(const ClauseUseRecord& usedClauses,
+                 TraceCheckProof&& tracecheck)
+    : d_inputClauseIds(), d_definitions(), d_tracecheck(tracecheck)
+{
+  // Step zero, save input clause Ids for future printing
+  std::transform(usedClauses.begin(),
+                 usedClauses.end(),
+                 std::back_inserter(d_inputClauseIds),
+                 [](const std::pair<ClauseId, prop::SatClause>& pair) {
+                   return pair.first;
+                 });
+
+  // Step one, verify the formula starts the proof
+  if (Configuration::isAssertionBuild())
+  {
+    for (size_t i = 0, n = usedClauses.size(); i < n; ++i)
+    {
+      Assert(d_tracecheck.d_lines[i].d_idx = i + 1);
+      Assert(d_tracecheck.d_lines[i].d_chain.size() == 0);
+      Assert(d_tracecheck.d_lines[i].d_clause.size()
+             == usedClauses[i].second.size());
+      for (size_t j = 0, m = usedClauses[i].second.size(); j < m; ++j)
+      {
+        Assert(usedClauses[i].second[j] == d_tracecheck.d_lines[i].d_clause[j]);
+      }
+    }
+  }
+
+  // Step two, identify definitions. They correspond to lines that follow the
+  // input lines, are in bounds, and have no justifying chain.
+  for (size_t i = usedClauses.size(), n = d_tracecheck.d_lines.size();
+       i < n && d_tracecheck.d_lines[i].d_chain.size() == 0;)
+  {
+    prop::SatClause c = d_tracecheck.d_lines[i].d_clause;
+    Assert(c.size() > 0);
+    Assert(!c[0].isNegated());
+
+    // Get the new variable of the definition -- the first variable of the
+    // first clause
+    prop::SatVariable newVar = c[0].getSatVariable();
+
+    // The rest of the literals in the clause of the 'other literals' of the def
+    std::vector<prop::SatLiteral> otherLiterals{++c.begin(), c.end()};
+
+    size_t nLinesForThisDef = 2 + otherLiterals.size();
+    // Look at the negation of the second literal in the second clause to get
+    // the old literal
+    d_definitions.emplace_back(newVar,
+                               ~d_tracecheck.d_lines[i + 1].d_clause[1],
```

Comment:
Could it ever happen that `i + 1` is out of bounds here? E.g. if a proof only has definitions for some weird reason?

Reply:
It turns out that definition actually manifest as `2 + n` lines in the TRACECHECK proof, where `n` is non-negative. Thus this line is still working with the same definition. That being said, if the definition was ill-formed (too short), this would be a segfault, and I should assert that the index is in bounds.

Reply:
Yes, it would be great to assert it because it is actually not guaranteed to segfault but could silently read garbage (hopefully this would be caught later though).

Reply:
:+1: